File: app/src/main/java/me/zhanghai/android/douya/network/api/info/frodo/Photo.java
/*
 * Copyright (c) 2017 Zhang Hai <Dreaming.in.Code.ZH@Gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.network.api.info.frodo;

import android.os.Parcel;

import com.google.gson.annotations.SerializedName;

import me.zhanghai.android.douya.ui.SizedImageItem;

public class Photo extends BaseItem implements SizedImageItem {

    @SerializedName("allow_comment")
    public boolean isCommentAllowed;

    public SimpleUser author;

    @SerializedName("create_time")
    public String creationTime;

    @SerializedName("comments_count")
    public int commentCount;

    public String description;

    public SizedImage image;

    @SerializedName("liked")
    public boolean isLiked;

    @SerializedName("likers_count")
    public int likerCount;

    @SerializedName("owner_uri")
    public String ownerUri;

    public int position;


    @Override
    public String getLargeUrl() {
        return image.getLargeUrl();
    }

    @Override
    public int getLargeWidth() {
        return image.getLargeWidth();
    }

    @Override
    public int getLargeHeight() {
        return image.getLargeHeight();
    }

    @Override
    public String getMediumUrl() {
        return image.getMediumUrl();
    }

    @Override
    public int getMediumWidth() {
        return image.getMediumWidth();
    }

    @Override
    public int getMediumHeight() {
        return image.getMediumHeight();
    }

    @Override
    public String getSmallUrl() {
        return image.getSmallUrl();
    }

    @Override
    public int getSmallWidth() {
        return image.getSmallWidth();
    }

    @Override
    public int getSmallHeight() {
        return image.getSmallHeight();
    }

    @Override
    public boolean isAnimated() {
        return image.isAnimated();
    }


    public static final Creator<Photo> CREATOR = new Creator<Photo>() {
        @Override
        public Photo createFromParcel(Parcel source) {
            return new Photo(source);
        }
        @Override
        public Photo[] newArray(int size) {
            return new Photo[size];
        }
    };

    public Photo() {}

    protected Photo(Parcel in) {
        super(in);

        isCommentAllowed = in.readByte() != 0;
        author = in.readParcelable(SimpleUser.class.getClassLoader());
        creationTime = in.readString();
        commentCount = in.readInt();
        description = in.readString();
        image = in.readParcelable(SizedImage.class.getClassLoader());
        isLiked = in.readByte() != 0;
        likerCount = in.readInt();
        ownerUri = in.readString();
        position = in.readInt();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);

        dest.writeByte(isCommentAllowed ? (byte) 1 : (byte) 0);
        dest.writeParcelable(author, flags);
        dest.writeString(creationTime);
        dest.writeInt(commentCount);
        dest.writeString(description);
        dest.writeParcelable(image, flags);
        dest.writeByte(isLiked ? (byte) 1 : (byte) 0);
        dest.writeInt(likerCount);
        dest.writeString(ownerUri);
        dest.writeInt(position);
    }
}

File: app/src/main/java/me/zhanghai/android/douya/api/info/Photo.kt
/*
 * Copyright (c) 2020 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.api.info

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class Photo(
    // BaseFeedableItem
    @Json(name = "abstract")
    override val abstractString: String = "",
    @Json(name = "url")
    override val alt: String = "",
    @Json(name = "cover_url")
    override val coverUrl: String = "",
    override val id: String = "",
    @Json(name = "sharing_url")
    override val sharingUrl: String = "",
    override val title: String = "",
    override val type: String = "",
    override val uri: String = "",

    val author: UserAbstract? = null,
    @Json(name = "collections_count")
    val collectionsCount: Int = 0,
    @Json(name = "comments_count")
    val commentsCount: Int = 0,
    @Json(name = "create_time")
    val createTime: String = "",
    val description: String = "",
    @Json(name = "donate_count")
    val donateCount: Int = 0,
    @Json(name = "allow_donate")
    val enableDonate: Boolean = false,
    val image: SizedImage? = null,
    @Json(name = "is_collected")
    val isCollected: Boolean = false,
    @Json(name = "is_original")
    val isOriginal: Boolean = false,
    @Json(name = "latest_comments")
    val latestComments: List<Comment> = emptyList(),
    val liked: Boolean = false,
    @Json(name = "likers_count")
    val likersCount: Int = 0,
    @Json(name = "allow_comment")
    val allowComment: Boolean = false,
    val owner: BaseFeedableItem? = null,
    @Json(name = "owner_uri")
    val ownerUri: String = "",
    val position: Int = 0,
    @Json(name = "reaction_type")
    val reactionType: Int = 0,
    @Json(name = "reactions_count")
    val reactionsCount: Int = 0,
    @Json(name = "read_count")
    val readCount: Int = 0,
    @Json(name = "reply_limit")
    val replyLimit: String = "",
    @Json(name = "reshares_count")
    val resharesCount: Int = 0,
    val status: Status? = null
) : IBaseFeedableItem
